### DISCUSSION TITLE
Audiobooks in Book library

### DISCUSSION THREAD

**Discussion by MaxTheSpy** [UnicornsOnLSD/finamp#990] [General]
Would it be possible to add functionality to stream audiobooks in finamp? The file formats are m4b. I use the book library because it keeps the listening progress.

**Comment by foss-** [comment]
Paging @Chaphasilor as suggested in https://github.com/jmshrv/finamp/issues/267#issuecomment-3000422859

Hijacking this existing discussion as the topic fits too well. For me it's only mp3s in the audiobook library which I use with the `Books` category.

When going to `Finamp > Settings > Select Music Libraries` the Music library is the only selectable library. So adding or using Audiobooks via Finamp is not an option currently.

If it it is possible to allow Books libraries and then ideally allow switching between music + books while remembering played state, that would be just fantastic 😇

**Reply by Chaphasilor** [reply]
okay so first @MaxTheSpy sorry for not responding to your question...

It should be possible to add basic support for Audiobooks (and audio playback of music videos), and I've implemented the very core of that on the [`redesign-more-libraries` branch](https://github.com/jmshrv/finamp/tree/redesign-more-libraries). Some more work is definitely needed.

Properly supporting things like authors and book series/collections would be more challenging though; I don't think we would add that any time soon, unless someone volunteers.

Is there any functionality that you'd consider absolutely necessary for audiobook libraries?

**Reply by foss-** [reply]
If you have a test build to play with that would be helpful to get a better understanding of where we are currently at.
To me anything that helps not losing track of played state would be highly welcome.
That is remembering playback position as otherwise it is just painful going through any longer duration files which are not possible to consume in a single go. And mark as played for albums (books?) which come in the form of multiple files would be helpful.

What I don't understand is the question about supporting authors. I use folder structure `Audiobooks\Author\Bookname`. If for a first iteration Author is just handled as a folder that would certainly be fine.

**Reply by Chaphasilor** [reply]
You can download a debug build from GitHub actions [here](https://github.com/jmshrv/finamp/actions/runs/16012162773). You'll have to be signed in to download (for some reason, don't ask me why).  
As you can see, right now I just added two new tabs (audio books and music videos) that fetch the respective item types. There's no support for showing or browsing authors at all (edit: turns out there kind of is, but that might be because I just used a music track for testing). That could be added by introducing a new list type type for audiobooks, a new menu, and another new tab for authors. I'm not sure how best to handle multi-part audiobooks, and I don't have a real audiobook library to test with. Jellyfin's public demo instance also doesn't offer any audio books.  
Maybe you could help out here? I'd just need a zip file with the correct folder structure and metadata to set up my own library for testing. The actual audio files can be just a few minutes of silence, or any non-copyrighted book or music track. It just needs to appear like an audio book.

For remembering playback position Finamp currently only stores previous queues, and can restore them manually. That also works for audio books out of the box. But the progress of a book won't be shown in the UI, nor will it resume from where you left if you start playing it directly.  
I think we could show playback progress and a "played" indicator relatively easily, especially if Jellyfin gives us the playback position in the user data (if not we can check if there's a restorable queue for the audio book).  
All in all though proper support for audio books will be quite a bit of work, as Finamp was never designed for that.

**Comment by MaxTheSpy** [comment]
For me personally when it comes to audiobooks the bare minimum is to be
able to listen and save progress and sort by author or series.

I've since swapped over to using audiobook shelf for listening and reading
so it's not a big concern anymore for me. But would be a nice to have all
my stuff in one place

**Reply by Chaphasilor** [reply]
Listening already works (in my development build), sorting should also be easy to add.  
Saving playback position theoretically also works, but is a bit cumbersome (you have to manually restore the queue).

I'll try to let you know if I have any updates!

**Comment by kontell** [comment]
I'd say the ability to adjust playback speed is a key feature. If you're used to listening at 1.5x, standard speed is a deal breaker.

Jellyfin does already remember resume position for audio files in a books library so it'd be good if however Finamp implements that feature it reads/ reports playback state to the server.

**Reply by Chaphasilor** [reply]
@kontell adjusting playback speed is already possible via the menu on the player screen. speed controls should automatically show up based on certain genre keywords and the duration of the track, but you can also force them to always be shown via the [🔗Playback Speed Visibility setting](https://intradeus.github.io/http-protocol-redirector?r=finamp://internal/settings/customization).

**Comment by BugZappa** [comment]
Bumping this thread, I was able to add my AudioBooks to a Playlist but Finamp refuses to play anything in that playlist even if it's downloaded locally.

**Reply by Chaphasilor** [reply]
@BugZappa so the audio books are in a Book library, yes?

**Reply by BugZappa** [reply]
> @BugZappa so the audio books are in a Book library, yes?

Yes on Jellyfin they're in the Books library, books/audiobooks can be placed in a Playlist, Finamp recognizes this but won't play anything in that playlist.

**Comment by n-buck** [comment]
May I ask what the status is here?
Is this merged into the redesign branch, already?
Do you need more people willing to test, or are you expecting some major upstream changes from jellyfin?
With jellyfin 12 I am able to listen to my audio books again (even tough organisation is still sub-optimal for multi-file audio-books) 🥳
